jetty-util/src/main/java/org/eclipse/jetty/util/Scanner.java

```diff
@@ -312,125 +312,111 @@ public int getScanInterval()
 
         File f = p.toFile();
         if (!f.exists() || f.isDirectory())
-            throw new IllegalStateException("Not file or doesn't exist: " + f.getCanonicalPath());
+            throw new IllegalStateException("Not file or doesn't exist: " + p);
 
-        try (AutoLock l = _lock.lock())
-        {
-            _scannables.put(p, null);
-        }
+        _scannables.putIfAbsent(p, null);
     }
 
     /**
      * Add a directory to be scanned. The directory must not be null and must exist.
      * 
      * @param p the directory to scan.
      * @return an IncludeExcludeSet to which the caller can add PathMatcher patterns to match
-     * @throws IOException
      */
-    public IncludeExcludeSet<PathMatcher, Path> addDirectory(Path p) throws IOException
+    public IncludeExcludeSet<PathMatcher, Path> addDirectory(Path p)
     {
+        if (isRunning())
+            throw new IllegalStateException("Scanner started");
+        
         if (p == null)
             throw new IllegalStateException("Null path");
 
         File f = p.toFile();
         if (!f.exists() || !f.isDirectory())
-            throw new IllegalStateException("Not directory or doesn't exist: " + f.getCanonicalPath());
+            throw new IllegalStateException("Not directory or doesn't exist: " + p);
 
-        try (AutoLock l = _lock.lock())
+        try
         {
-            IncludeExcludeSet<PathMatcher, Path> includesExcludes = _scannables.get(p);
-            if (includesExcludes == null)
-            {
-                includesExcludes = new IncludeExcludeSet<>(PathMatcherSet.class);
-                _scannables.put(p.toRealPath(), includesExcludes);
-            }
+            Path real = p.toRealPath();
+            IncludeExcludeSet<PathMatcher, Path> includesExcludes = new IncludeExcludeSet<>(PathMatcherSet.class);
+            IncludeExcludeSet<PathMatcher, Path> prev = _scannables.putIfAbsent(real, includesExcludes);
+            if (prev != null)
+                includesExcludes = prev;
             return includesExcludes;
         }
-    }
-
-    @Deprecated
-    public List<File> getScanDirs()
-    {
-        ArrayList<File> files = new ArrayList<>();
-        for (Path p : _scannables.keySet())
-            files.add(p.toFile());
-        return Collections.unmodifiableList(files);
+        catch (IOException e)
+        {
+            throw new IllegalStateException(e);
+        }
     }
     
-    public Set<Path> getScannables()
-    {
-        return _scannables.keySet();
-    }
 
     /**
-     * @param recursive True if scanning is recursive
-     * @see #setScanDepth(int)
+     * Apply a filter to files found in the scan directory.
+     * Only files matching the filter will be reported as added/changed/removed.
+     *
+     * @param filter the filename filter to use
      */
     @Deprecated
-    public void setRecursive(boolean recursive)
+    public void setFilenameFilter(FilenameFilter filter)
```

Comment:
Still worth to keep around `Deprecated` method?

Reply:
I don't really want to mess with the WebAppProvider's filter for finding webapps to deploy at this late stage - it can be refactored at a later stage, so I've raised #5748 to track it.